File: 99-Continue/162/RightTraversal.java
import java.util.*;

public class RightTraversal {

    public static void main(String[] args) {
       Scanner sc = new Scanner(System.in);
       int n = sc.nextInt();
       sc.nextLine();
       String s = sc.nextLine();
       String[] strarr = s.split(" ");
       BinaryTree tree = new BinaryTree();
       tree.buildTree(strarr);
       tree.rightTraversal();
       sc.close();
    }
}

class Node{
    int data;
    Node left, right;
    Node(int data){
        this.data = data;
        this.left = this.right = null;
    }
}

class BinaryTree{
    Node root;
    BinaryTree(){
        this.root = null;
    }
    
    Node buildTree(String[] arr){
        if(arr.length==0){
            return root;
        }
        
        Queue<Node> queue = new LinkedList<>();
        root = new Node(Integer.parseInt(arr[0]));
        queue.offer(root);
        int i=1;
        while(!queue.isEmpty()&&i<arr.length){
            Node current = queue.poll();
            if(i<arr.length && !arr[i].equals("null")){
                current.left = new Node(Integer.parseInt(arr[i]));
                queue.offer(current.left);
            }
            i++;
            if(i<arr.length && !arr[i].equals("null")){
                current.right = new Node(Integer.parseInt(arr[i]));
                queue.offer(current.right);
            }
            i++;
            
        }
        return root;
    }
    
    void rightTraversal(){
        if(root==null){
            return;
        }
        
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        System.out.print(root.data+" ");
        while(!queue.isEmpty()){
            int size=queue.size();
            int val = -1;
            while(size-->0){
                Node current = queue.poll();
                if(current.left!=null){
                    queue.offer(current.left);
                    val = current.left.data;
                }
                if(current.right!=null){
                    queue.offer(current.right);
                    val=current.right.data;
                }
            }
            if(val!=-1){
                System.out.print(val+" ");
            }
            
        }
    }
}

File: 99-Continue/162/RightTraversal.py
from collections import deque

class Node:
    def __init__(self,data:int):
        self.data = data
        self.left = self.right = None

class BinaryTree:
    def __init__(self):
        self.root = None
        
    def buildTree(self,arr:list[str])->None:
        if len(arr)==0:
            return self.root
        
        self.root=Node(int(arr[0]))
        queue = deque([self.root])
        
        i=1
        while queue and i<len(arr):
            current = queue.popleft()
            if i<len(arr) and arr[i]!="null":
                current.left = Node(int(arr[i]))
                queue.append(current.left)
            i+=1
            if i<len(arr) and arr[i]!="null":
                current.right = Node(int(arr[i]))
                queue.append(current.right)
            i+=1
        
        return Node
    
    def rightTraversal(self):
        if self.root is None:
            return None
        
        queue = deque([self.root])
        print(self.root.data,end=" ")
        while queue:
            size = len(queue)
            val = -1
            while size>0:
                current = queue.popleft()
                if current.left is not None:
                    queue.append(current.left)
                    val=current.left.data
                if current.right is not None:
                    queue.append(current.right)
                    val=current.right.data
                size-=1
            if val!=-1:
                print(val,end=" ")

def main():
    int(input())
    tree = BinaryTree()
    s = input().split()
    tree.buildTree(s)
    tree.rightTraversal()
    
if __name__ == "__main__":
    main()
    
        
            
        
        
        
    